doc/cephfs/capabilities.rst

```diff
@@ -110,3 +110,70 @@ capabilities. For example:
 The 'p' represents the pin. Each capital letter corresponds to the shift
 values, and the lowercase letters after each shift are for the actual
 capabilities granted in each shift.
+
+The relation between the lock states and the capabilities
+---------------------------------------------------------
+In MDS there are four different locks for each inode, they are simplelock,
+scatterlock, filelock and locallock. Each lock has several different lock
+states, and the MDS will issue capabilities to clients based on the lock
+state.
+
+In each state the MDS Locker will always try to issue all the capabilities to the
+clients allowed, even some capabilities are not needed or wanted by the clients,
+as pre-issuing capabilities could reduce latency in some cases.
+
+If there is only one client, usually it will be the loner client for all the inodes.
+While in multiple clients case, the MDS will try to caculate a loner client out for
+each inode depending on the capabilities the clients (needed | wanted), but usually
+it will fail. The loner client will always get all the capabilities.
+
+The filelock will control files' partial metadatas' and the file contents' access
+permissions. The metadatas include **mtime**, **atime**, **size**, etc.
+
+**Fs**: Once a client has it, all other clients are denied **Fw**.
+
+**Fx**: Only the loner client is allowed this capability. Once the lock state transitions
+        to LOCK_EXCL, the loner client is granted this along with all other file capabilities
+        except the **Fl**.
+
+**Fr**: Once a client has it, the **Fb** capability will be already revoked from all the
+        other clients.
+
+        If all the clients will only request to read the same file, the lock state will be
+        transferred to LOCK_SYNC stable state finally. And all the client could have the
+        **Fscrl** capabilities from the auth MDS and **Fscr** capabilities from the
+        replica MDSes.
+
+        If there is any client wants to write the file at the same time, then the lock
+        state will be transferred to LOCK_MIX stable state finally and all the client
+        could have the **Frwl** capabilities from the auth MDS, and the **Fr** from the
+        replica MDSes. The **Fcb** capabilities will be can't have from all the clients and
+        the clients will do sync read/write.
```

Comment:
I'm not sure I understand how the above relates to **Fr** caps. Also:

"If there is any client wants to write the file at the same time"

How does the MDS know that the client wants to write? I assume by requesting **Fw** caps?

Reply:
> I'm not sure I understand how the above relates to **Fr** caps. Also:
> 

The above will be the same in `**Fw**` section below. And I just copied it to there.

> "If there is any client wants to write the file at the same time"
>

`If there is any client wants to write the file at the same time` 

This is possibly a bit confusing, it should be `If there are clients want to read and write the file at the same time`.
 
> How does the MDS know that the client wants to write? I assume by requesting **Fw** caps?

Yeah, right. You know here the `write/read` will be interpreted to `Fw/Fr` caps in clients.

The `MDS` will record all the clients' `wanted` caps, such as when clients do `open(..., O_RDWR)`, they will request `Fw` caps via `wanted`. And the `MDS` will calculate and transfer to the correct `filelock` state by depending on the `wanted` caps.